internal/engine/wazevo/backend/isa/arm64/instr_encoding.go

```diff
@@ -865,33 +929,64 @@ func encodeMoveFromVec(rd, rn uint32, arr vecArrangement, index vecIndex) uint32
 	default:
 		panic("Unsupported arrangement " + arr.String())
 	}
-	return 0b0_001110000<<21 | q<<30 | imm5<<16 | 0b001111<<10 | rn<<5 | rd
+	if signed {
+		op, imm4 = 0, 0b0101
+	} else {
+		op, imm4 = 0, 0b0111
+	}
+	return op<<29 | 0b01110000<<21 | q<<30 | imm5<<16 | imm4<<11 | 1<<10 | rn<<5 | rd
 }
 
-// encodeVecDup encodes as "Duplicate general-purpose register to vector."
+// encodeVecDup encodes as "Duplicate general-purpose register to vector" DUP (general)
+// and "Duplicate vector element to vector or scalar" DUP (element).
 // (represented as `dup`)
 // https://developer.arm.com/documentation/ddi0596/2020-12/SIMD-FP-Instructions/DUP--general---Duplicate-general-purpose-register-to-vector-?lang=en
-func encodeVecDup(rd, rn uint32, arr vecArrangement) uint32 {
-	var q, imm5 uint32
-	switch arr {
-	case vecArrangement8B:
-		q, imm5 = 0b0, 0b1
-	case vecArrangement16B:
-		q, imm5 = 0b1, 0b1
-	case vecArrangement4H:
-		q, imm5 = 0b0, 0b10
-	case vecArrangement8H:
-		q, imm5 = 0b1, 0b10
-	case vecArrangement2S:
-		q, imm5 = 0b0, 0b100
-	case vecArrangement4S:
-		q, imm5 = 0b1, 0b100
-	case vecArrangement2D:
-		q, imm5 = 0b1, 0b1000
-	default:
-		panic("Unsupported arrangement " + arr.String())
+// https://developer.arm.com/documentation/ddi0596/2020-12/SIMD-FP-Instructions/DUP--element---Duplicate-vector-element-to-vector-or-scalar-
+func encodeVecDup(rd, rn uint32, arr vecArrangement, srcIndex vecIndex) uint32 {
+	var op, q, imm5 uint32
+	if srcIndex != vecIndexNone {
+		op = 0b1
+		q = 0b1
+		switch arr {
+		case vecArrangementB:
+			imm5 |= 0b1
+			imm5 |= uint32(srcIndex) << 1
+		case vecArrangementH:
+			imm5 |= 0b10
+			imm5 |= uint32(srcIndex) << 2
+		case vecArrangementS:
+			imm5 |= 0b100
+			imm5 |= uint32(srcIndex) << 3
+		case vecArrangementD:
+			imm5 |= 0b1000
+			imm5 |= uint32(srcIndex) << 4
+		default:
+			panic("unsupported arrangement" + arr.String())
+		}
```

Comment:
it's probably better to add vecDupElement with its own encoding... there is not much overlap